WhateverGreen/kern_igfx.cpp

```diff
@@ -651,6 +721,585 @@ int IGFX::wrapReadAUX(void *that, IORegistryEntry *framebuffer, uint32_t address
+		return 0;
+	}
+	params->pdiv = context.pdiv;
+	params->qdiv = context.qdiv;
+	params->kdiv = context.kdiv;
+	params->multiplier = multiplier;
+	params->fraction = fraction;
+	params->cf15625 = cf15625;
+	DBGLOG("igfx", "SC: ComputeHdmiP0P1P2() DInfo: CTRC parameters have been populated successfully.");
+	return 0;
+}
+
+IGFX::LSPCON::LSPCON(void *controller, IORegistryEntry *framebuffer, void *displayPath) {
+	this->controller = controller;
+	this->framebuffer = framebuffer;
+	this->displayPath = displayPath;
+	this->isActive = false;
+	this->index = OSDynamicCast(OSNumber, framebuffer->getProperty("IOFBDependentIndex"))->unsigned32BitValue();
+}
+
+IOReturn IGFX::LSPCON::probe() {
+	// Read the adapter info
+	uint8_t buffer[128];
+	bzero(buffer, 128);
+	IOReturn retVal = callbackIGFX->advReadI2COverAUX(controller, framebuffer, displayPath, DP_DUAL_MODE_ADAPTER_I2C_ADDR, 0x00, 128, buffer, 0);
+	if (retVal != kIOReturnSuccess) {
+		SYSLOG("igfx", "SC: LSPCON::probe() Error: [FB%d] Failed to read the LSPCON adapter info. RV = 0x%llx.", index, retVal);
+		return retVal;
+	}
+	
+	// Start to parse the adapter info
+	auto info = reinterpret_cast<DisplayPortDualModeAdapterInfo*>(buffer);
+	// Guard: Check whether this is a LSPCON adapter
+	if (!isLSPCONAdapter(info)) {
+		SYSLOG("igfx", "SC: LSPCON::probe() Error: [FB%d] Not a LSPCON DP-HDMI adapter. AdapterID = 0x%02x.", index, info->adapterID);
+		return kIOReturnNotFound;
+	}
+	
+	// Parse the chip vendor
+	char device[8];
+	bzero(device, 8);
+	lilu_os_memcpy(device, info->deviceID, 6);
+	DBGLOG("igfx", "SC: LSPCON::probe() DInfo: [FB%d] Found the LSPCON adapter: %s %s.", index, getVendorString(parseVendor(info)), device);
+	
+	// Parse the current adapter mode
+	Mode mode = parseMode(info->lspconCurrentMode);
+	DBGLOG("igfx", "SC: LSPCON::probe() DInfo: [FB%d] The current adapter mode is %s.", index, getModeString(mode));
+	switch (mode) {
+		case Mode::LevelShifter:
+			break;
+			
+		case Mode::ProtocolConverter:
+			isActive = true;
+			break;
+			
+		default:
+			SYSLOG("igfx", "SC: LSPCON::probe() Error: [FB%d] Cannot detect the current adapter mode. Assuming Level Shifter mode.", index);
+			break;
+	}
+	
+	return kIOReturnSuccess;
+}
+
+IOReturn IGFX::LSPCON::getMode(Mode *mode) {
+	uint8_t byte;
+	IOReturn retVal = kIOReturnAborted;
+	
+	// Guard: The given `mode` pointer cannot be NULL
+	if (mode != nullptr) {
+		// Try at most 5 times
+		for (int attempt = 0; attempt < 5; attempt++) {
+			// Read from the adapter @ 0x40; offset = 0x41
+			retVal = callbackIGFX->advReadI2COverAUX(controller, framebuffer, displayPath, DP_DUAL_MODE_ADAPTER_I2C_ADDR, DP_DUAL_MODE_LSPCON_CURRENT_MODE, 1, &byte, 0);
+			
+			// Guard: Can read the current adapter mode successfully
+			if (retVal == kIOReturnSuccess) {
+				DBGLOG("igfx", "SC: LSPCON::getMode() DInfo: [FB%d] The current mode value is 0x%02x.", index, byte);
+				*mode = parseMode(byte);
+				break;
+			}
+			
+			// Sleep 1 ms
```

Comment:
Please explain why you need to sleep for 1 ms. It is relatively clear that IOSleep(1) does that. What is not clear, is why it is needed. Perhaps this comment needs to be removed entirely, and `Try at most 5 times` needs to be more descriptive.